### PR TITLE
[compiler] Add JSX inlining optimization

### PR DESCRIPTION
**Pull request by jackpope**
This adds an `InlineJsxTransform` optimization pass, toggled by the `enableInlineJsxTransform` flag. When enabled, JSX will be transformed into React Element object literals, preventing runtime overhead during element creation.

TODO:
- [ ] Add conditionals to make transform PROD-only
- [ ] Make the React element symbol configurable so this works with runtimes that support `react.element` or `react.transitional.element`
- [ ] Look into additional optimization to pass props spread through directly if none of the properties are mutated

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/Optimization/InlineJsxTransform.ts [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+       * Scope ranges should always align to start at the 'scope' terminal
+       * and end at the first instruction of the fallthrough block
+       */
+      const fallthroughBlock = fn.body.blocks.get(terminal.fallthrough)!;
+      const firstId =
+        fallthroughBlock.instructions[0]?.id ?? fallthroughBlock.terminal.id;
+      terminal.scope.range.start = terminal.id;
+      terminal.scope.range.end = firstId;
+    }
+  }
```
Comment (jackpope) [inline compiler/packages/babel-plugin-react-compiler/src/Optimization/InlineJsxTransform.ts:402]: I stole this from BuildReactiveScopeTerminalsHIR.ts. Should this be a shared utility?

**Comment by sebmarkbage** [comment]
This should not be shipped in code that's compiled for npm since it is not compatible with multiple React versions.

It's also important that it doesn't ship in an RSC layer which needs a runtime to do the warm up pass.

(In addition, previous testing showed that this is not actually better than a minimal runtime in most environments due to the additional compilation cost.)

**Comment by sebmarkbage** [comment]
To clarify the last statement.

Most testing of inline objects have compared it against the React runtime which has overhead mainly due to legacy features like the refs and the backwards compat with defaultProps. A proper A/B test should test the performance of inlining compared to an optimized runtime function that does the same thing that the inlining would do. Hopefully in React 20+ that can just be the same thing in the plain `React.jsx` function assuming deprecations land.

In that comparison it was previous discovered that the compilation time for VMs like V8 that can JIT a callsite to determine which hidden class to create but figuring that out for each callsite means it has to transition through the hidden class discovery like `object with $$typeof -> object with type -> object with key -> ...`. Which significantly slowed down initialization compared to just making it a call and then that one call getting in a JIT:ed callsite.

This effect wasn't as big in JSC possibly due to faster compiler but also due to lack of other call optimizations not being as good so the relative size isn't necessarily there.

For Hermes, since the compilation time is entirely offline the cost of inlining isn't as big so the runtime overhead of the extra call might make it worth it there but it comes as the cost of larger byte code which may or may not matter.

Basically I'm skeptical overall it is actually good to inline in most circumstances and it would be better to just have a single shared runtime. Given that the output is not multi-version compatible and it's actually worst in the largest platform (Chrome). And it's not compatible with JSX prewarming in RSC.

**Comment by jackpope** [comment]
@sebmarkbage For now this optimization is meant to be turned on for Hermes only. I benchmarked and profiled the inlining on some test apps and saw some regressions in V8 as you described. In some apps interactions were faster, but memory regressed significantly. Running with ` --jitless` was more stable so that lines up with your observations around JIT issues.

With Hermes, inlining showed wins to interaction performance and memory, even compared to a React 20 style jsx function. With this implementation we can test in production against real apps and understand the full impact of the change and if its worth maintaining the optimization for Hermes only.

File: compiler/packages/babel-plugin-react-compiler/src/Optimization/InlineJsxTransform.ts (hunk 1 above)
Comment (josephsavona) [reply compiler/packages/babel-plugin-react-compiler/src/Optimization/InlineJsxTransform.ts:402]: yeah!

File: compiler/packages/babel-plugin-react-compiler/src/Optimization/InlineJsxTransform.ts [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+  };
+  nextInstructions.push(symbolInstruction);
+
+  const symbolForPlace = createTemporaryPlace(fn.env, instr.value.loc);
+  const symbolForInstruction: Instruction = {
+    id: makeInstructionId(0),
+    lvalue: {...symbolForPlace, effect: Effect.Read},
+    value: {
+      kind: 'PropertyLoad',
+      object: symbolInstruction.lvalue,
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/Optimization/InlineJsxTransform.ts:52]: note that places always have to be copied, not used by reference, so `{...symbolPlace}` here

File: compiler/packages/babel-plugin-react-compiler/src/Optimization/InlineJsxTransform.ts [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+function createTagProperty(
+  fn: HIRFunction,
+  instr: Instruction,
+  nextInstructions: Array<Instruction>,
+  componentTag: BuiltinTag | Place,
+): ObjectProperty {
+  let tagProperty: ObjectProperty;
+  switch (componentTag.kind) {
+    case 'BuiltinTag':
+      const tagPropertyPlace = createTemporaryPlace(fn.env, instr.value.loc);
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/Optimization/InlineJsxTransform.ts:105]: always wrap cases in `{}`

**Review by josephsavona — CHANGES_REQUESTED** [review]
This is great!  A few minor changes and then we should be able to land and experiment.

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/inline-jsx-transform.expect.md [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  const $ = _c2(2);
+  const { children, ref } = t0;
+  let t1;
+  if ($[0] !== children) {
+    t1 = {
+      $$typeof: Symbol.for("react.transitional.element"),
+      type: "div",
+      ref: ref,
+      key: null,
+      props: { children: [children] },
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/inline-jsx-transform.expect.md:66]: should this be an array (`[children]`) or just the raw `children` value?

**Review by josephsavona — APPROVED** [review]
Looks great! Just one quick thing to double-check for when there's a single child element.

**Comment by josephsavona** [comment]
Nice! Let's ship and iterate
